components/uniset/src/enum_props.rs

```diff
@@ -15,50 +13,85 @@ pub enum EnumeratedProperty {
-    PrivateUse = 17,
-    Surrogate = 18,
-    LowercaseLetter = 2,
-    ModifierLetter = 4,
-    OtherLetter = 5,
-    TitlecaseLetter = 3,
-    UppercaseLetter = 1,
-    SpacingMark = 8,
-    EnclosingMark = 7,
-    NonspacingMark = 6,
-    Digit = 9,
-    LetterNumber = 10,
-    OtherNumber = 11,
-    ConnectorPunctuation = 22,
-    DashPunctuation = 19,
-    ClosePunctuation = 21,
-    FinalPunctuation = 29,
-    InitialPunctuation = 28,
-    OtherPunctuation = 23,
-    OpenPunctuation = 20,
-    CurrencySymbol = 25,
-    ModifierSymbol = 26,
-    MathSymbol = 24,
-    OtherSymbol = 27,
-    LineSeparator = 13,
-    ParagraphSeparator = 14,
-    SpaceSeparator = 12,
+
+    UppercaseLetter = 1 << 1,
+    LowercaseLetter = 1 << 2,
+    TitlecaseLetter = 1 << 3,
+    ModifierLetter = 1 << 4,
+    OtherLetter = 1 << 5,
+    CasedLetter =
+        Self::UppercaseLetter as u32 | Self::LowercaseLetter as u32 | Self::TitlecaseLetter as u32,
+    Letter = Self::CasedLetter as u32 | Self::ModifierLetter as u32 | Self::OtherLetter as u32,
+
+    NonspacingMark = 1 << 6,
+    EnclosingMark = 1 << 7,
+    SpacingMark = 1 << 8,
+    Mark = Self::NonspacingMark as u32 | Self::EnclosingMark as u32 | Self::SpacingMark as u32,
+
+    Digit = 1 << 9,
+    LetterNumber = 1 << 10,
+    OtherNumber = 1 << 11,
+    Number = Self::Digit as u32 | Self::LetterNumber as u32 | Self::OtherNumber as u32,
+
+    SpaceSeparator = 1 << 12,
+    LineSeparator = 1 << 13,
+    ParagraphSeparator = 1 << 14,
+    Separator =
+        Self::SpaceSeparator as u32 | Self::LineSeparator as u32 | Self::ParagraphSeparator as u32,
+
+    Control = 1 << 15,
+    Format = 1 << 16,
+    PrivateUse = 1 << 17,
+    Surrogate = 1 << 18,
+    Other = Self::Control as u32
+        | Self::Format as u32
+        | Self::PrivateUse as u32
+        | Self::Surrogate as u32,
+
+    DashPunctuation = 1 << 19,
+    OpenPunctuation = 1 << 20,
+    ClosePunctuation = 1 << 21,
+    ConnectorPunctuation = 1 << 22,
+    OtherPunctuation = 1 << 23,
+    InitialPunctuation = 1 << 28,
+    FinalPunctuation = 1 << 29,
+    Punctuation = Self::DashPunctuation as u32
+        | Self::OpenPunctuation as u32
+        | Self::ClosePunctuation as u32
+        | Self::ConnectorPunctuation as u32
+        | Self::OtherPunctuation as u32
+        | Self::InitialPunctuation as u32
+        | Self::FinalPunctuation as u32,
+
+    MathSymbol = 1 << 24,
+    CurrencySymbol = 1 << 25,
+    ModifierSymbol = 1 << 26,
+    OtherSymbol = 1 << 27,
+    Symbol = Self::MathSymbol as u32
+        | Self::CurrencySymbol as u32
+        | Self::ModifierSymbol as u32
+        | Self::OtherSymbol as u32,
 }
 
 //// Enumerated property Script.
 ///
 /// For more information, see UAX #24: http://www.unicode.org/reports/tr24/.
 /// See UScriptCode in ICU4C.
+///
+/// This enum only contains variants for scripts that are used in the Unicode
+/// Property Database.
 #[derive(Clone, PartialEq, Debug)]
 pub enum Script {
```

Comment:
In the context of uniset, we already fail relatively gracefully if the UCD changes. You can't request the set of characters in a newly-added script until we add support, but the datagen should work out fine if somebody downloads a newer version of the UCD. I can imagine that it might be more of an issue for the code point trie, but I think it's best to address that as part of the code point trie work where we have better perspective.

We should probably add `#[non_exhaustive]` here, though. (And maybe also GeneralCategory? I thought I had read somewhere that ICU had committed to not adding new general categories, but I can't find any evidence.)